src/core/include/mp-units/bits/fixed_point.h

```diff
@@ -0,0 +1,309 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+// IWYU pragma: private, include <mp-units/framework.h>
+#include <mp-units/framework/magnitude.h>
+
+#ifndef MP_UNITS_IN_MODULE_INTERFACE
+#ifdef MP_UNITS_IMPORT_STD
+import std;
+#else
+#include <bit>
+#include <concepts>
+#include <cstdint>
+#include <cstdlib>
+#include <limits>
+#include <numbers>
+#endif
+#endif
+
+namespace mp_units {
+namespace detail {
+
+// this class synthesizes a double-width integer from two base-width integers.
+template<std::integral T>
+struct double_width_int {
+  static constexpr bool is_signed = std::is_signed_v<T>;
+  static constexpr std::size_t base_width = std::numeric_limits<std::make_unsigned_t<T>>::digits;
+  static constexpr std::size_t width = 2 * base_width;
+
+  using Th = T;
+  using Tl = std::make_unsigned_t<T>;
+
+  constexpr double_width_int() = default;
+  constexpr double_width_int(const double_width_int&) = default;
+
+  constexpr double_width_int& operator=(const double_width_int&) = default;
+
+  constexpr double_width_int(Th hi_, Tl lo_) : hi(hi_), lo(lo_) {}
+
+  explicit(true) constexpr double_width_int(long double v)
+  {
+    constexpr auto scale = int_power<long double>(2, base_width);
+    constexpr auto iscale = 1.l / scale;
+    auto scaled = v * iscale;
+    hi = static_cast<Th>(scaled);
+    auto resid = (scaled - static_cast<long double>(hi));
+    if (resid < 0) {
+      --hi;
+      resid += 1;
+    }
+    lo = static_cast<Tl>(resid * scale);
+  }
+  template<std::integral U>
+    requires(is_signed || !std::is_signed_v<U>)
+  explicit(false) constexpr double_width_int(U v)
+  {
+    if constexpr (is_signed) {
+      hi = v < 0 ? Th{-1} : Th{0};
+    } else {
+      hi = 0;
+    }
+    lo = static_cast<Tl>(v);
+  }
+
+  explicit(true) constexpr operator Th() const { return static_cast<Th>(lo); }
+
+  constexpr auto operator<=>(const double_width_int&) const = default;
+
+  // calculates the double-width product of two base-size integers; this implementation requires at least one of them to
+  // be unsigned
+  static constexpr double_width_int wide_product_of(Th lhs, Tl rhs)
```

Comment:
Is this a factory function? If so, shouldn't the constructors be `private`?

Reply:
This is a factory function, but note that it is not equivalent to the constructor taking a `(hi, lo)` pair. `wide_product_of(lhs,rhs)` creates an instance representing `lhs * rhs`, while `double_width_int(hi,lo)` creates an instance representing `(hi<<base_width) + lo`.

One could argue though that there is a certain risk of confusion about the semantics of the two-argument constructor, so we may want to still make it private and instead expose a factory function `from_nibbles` or similar.